src/rust/cryptography-x509-validation/src/policy/mod.rs

```diff
@@ -250,15 +279,315 @@ impl<'a, B: CryptoOps> Policy<'a, B> {
+            ext_policy.permits(self, cert, extensions)?;
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether the given EE certificate is compatible with this policy.
+    pub(crate) fn permits_ee(
+        &self,
+        cert: &Certificate<'_>,
+        extensions: &Extensions<'_>,
+    ) -> Result<(), PolicyError> {
+        self.permits_basic(cert)?;
+
+        for ext_policy in self.ee_extension_policies.iter() {
+            ext_policy.permits(self, cert, extensions)?;
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether `issuer` is a valid issuing CA for `child` at a
+    /// path-building depth of `current_depth`.
+    ///
+    /// This checks that `issuer` is permitted under this policy and that
+    /// it was used to sign for `child`.
+    ///
+    /// On success, this function returns the new path-building depth. This
+    /// may or may not be a higher number than the original depth, depending
+    /// on the kind of validation performed (e.g., whether the issuer was
+    /// self-issued).
+    pub(crate) fn valid_issuer(
+        &self,
+        issuer: &Certificate<'_>,
+        child: &Certificate<'_>,
+        current_depth: u8,
+        issuer_extensions: &Extensions<'_>,
+    ) -> Result<u8, PolicyError> {
+        // The issuer needs to be a valid CA at the current depth.
+        self.permits_ca(issuer, current_depth, issuer_extensions)?;
+
+        // CA/B 7.1.3.1 SubjectPublicKeyInfo
+        if !self
+            .permitted_public_key_algorithms
+            .contains(&child.tbs_cert.spki.algorithm)
+        {
+            // TODO: Should probably include the OID here.
+            return Err("Forbidden public key algorithm".into());
+        }
+
+        // CA/B 7.1.3.2 Signature AlgorithmIdentifier
+        if !self
+            .permitted_signature_algorithms
+            .contains(&child.signature_alg)
+        {
+            // TODO: Should probably include the OID here.
+            return Err("Forbidden signature algorithm".into());
+        }
+
+        let pk = self
+            .ops
+            .public_key(issuer)
+            .map_err(|_| PolicyError::Other("issuer has malformed public key"))?;
+        if self.ops.verify_signed_by(child, pk).is_err() {
+            return Err(PolicyError::Other("signature does not match"));
+        }
+
+        // Self-issued issuers don't increase the working depth.
+        match cert_is_self_issued(issuer) {
+            true => Ok(current_depth),
+            false => Ok(current_depth
+                .checked_add(1)
+                .ok_or(PolicyError::Other("current depth calculation overflowed"))?),
         }
     }
 }
 
+fn valid_validity_date(validity_date: &Time) -> Result<(), PolicyError> {
+    const GENERALIZED_DATE_CUTOFF_YEAR: u16 = 2050;
+    match validity_date {
+        Time::UtcTime(_) => {
+            // NOTE: The `asn1::UtcTime` constructor already checks the underlying datetime year so
```

Comment:
This is a bit odd. I had to remove this logic since it's impossible to exercise.

I wonder if it's possible to add similar checks to `asn1::GeneralizedTime` and remove this validation function entirely. I'll grep around and have a look to see where else it's used.

Reply:
I don't believe it's possible to put a check there: the restrictions in which dates it can be used for are specific to x.509.